pkg/neonvm/controllers/vm_cert_controller.go

```diff
@@ -0,0 +1,341 @@
+package controllers
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/cert-manager/cert-manager/pkg/apis/certmanager"
+	certv1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	cmmeta "github.com/cert-manager/cert-manager/pkg/apis/meta/v1"
+	"github.com/cert-manager/cert-manager/pkg/util/pki"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	vmv1 "github.com/neondatabase/autoscaling/neonvm/apis/neonvm/v1"
+)
+
+func (r *VMReconciler) reconcileCertificateSecret(ctx context.Context, vm *vmv1.VirtualMachine) (*corev1.Secret, error) {
+	log := log.FromContext(ctx)
+
+	certSecret := &corev1.Secret{}
+
+	// Check if the TLS secret exists, if not start the creation routine.
+	err := r.Get(ctx, types.NamespacedName{Name: vm.Status.TLSSecretName, Namespace: vm.Namespace}, certSecret)
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "Failed to get vm-runner TLS secret")
+		return nil, err
+	}
+
+	notfound := false
+	if err != nil /* not found */ {
+		msg := fmt.Sprintf("VirtualMachine %s TLS secret %s not found", vm.Name, vm.Status.TLSSecretName)
+		r.Recorder.Event(vm, "Normal", "SigningCertificate", msg)
+
+		notfound = true
+	} else if time.Now().After(vm.Status.TLSRenewAt.Time) /* due for renewal */ {
+		msg := fmt.Sprintf("VirtualMachine %s TLS secret %s is due for renewal", vm.Name, vm.Status.TLSSecretName)
+		r.Recorder.Event(vm, "Normal", "SigningCertificate", msg)
+	} else /* certificate exists and does not need renewal */ {
+		return certSecret, nil
+	}
+
+	// Check if the TLS private key temporary secret exists, if not create a new one
+	tmpKeySecret := &corev1.Secret{}
+	err = r.Get(ctx, types.NamespacedName{Name: fmt.Sprintf("%s-tmp", vm.Status.TLSSecretName), Namespace: vm.Namespace}, tmpKeySecret)
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "Failed to get vm-runner TLS secret")
+		return nil, err
+	} else if err != nil /* not found */ {
+		tmpKeySecret, err = r.createTlsTmpSecret(ctx, vm)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	key, err := pki.DecodePrivateKeyBytes(tmpKeySecret.Data[corev1.TLSPrivateKeyKey])
+	if err != nil {
+		log.Error(err, "Failed to decode TLS private key")
```

Comment:
question(blocking): is not returning here intentional? If so, what do we expect to happen by falling through?

Reply:
now returning an error